(145) Send room in description

dgmstuart opened the pull request:
- Introduce 'mandatory' questions
- We talked about sending it as a special field, but this doesn't seem to have been locked in in the API standards, so send it in the description instead

dgmstuart committed:
(refactor) move QuestionsValidator into a class

We're going to extend this so it would be useful to be able to test it
more quickly and directly.

- Keep the QuestionSet specs as integration tests for the existing
  checks.
- Backfill some specific test cases around successes

dgmstuart committed:
(feature) Mandatory questions can be specified

The questions.yml file will be editable by non-developers, but we rely
on specific questions being available - e.g. the first page. We
therefore need to be able to validate that these question ids exist at
question load time.

dgmstuart committed:
(refactor) Extract a QuestionSetLoader

Adding a new parameter to QuestionsValidator makes the QuestionSet
intializer too long, so this feels like a good opportunity to split out
a loader class, which is a step in the direction of loading questions
once and accessing them in many places.

Note that the fact that `questions` are passed as an initialization
parameter to QuestionsValidator prevents us from injecting that class as
well. Not too bothered by that: we can do that refactoring in the future
if it seems useful.

dgmstuart committed:
(feature) fail if there is no 'location' question

We use this as the first question, so it has to exist. Some of the
existing tests used a questions file which didn't have this question.

dgmstuart committed:
(delete) remove unused accessor

Not sure how this wasn't caught by code coverage - it's certainly not
used anywhere.

dgmstuart committed:
(feature) Questions have IDs

Previously the ID was just the key used to refer to a question. Now we
need to know what the ID of the question was, and passing it as part of
the Question object feels better than passing that as an additional
parameter from the controller.

Unfortunately this requires changing a lot of specs...

dgmstuart committed:
(feature) Room is stored and shown on confirmation

Since we're depending on this field being there, we need to make it
mandatory and update fixtures accordingly.

dgmstuart committed:
(feature) Send room in description

If room is specified, append it to the description when creating the
repair

At some point we said we would send this as a separate field, but it
looks like we didn't specify this in the API standards, and I'm not sure
it's a good idea anyway, since we wouldn't necessarily be able to get it
back reliably (assuming it would be mashed into the description field by
the API anyway).

dgmstuart committed:
(fix) Don't display room for communal areas

We didn't have any specs for non-internal spaces. Not sure if this one
is good enough.

I don't like having a conditional in the view. Perhaps it would be
better to have a single 'location' field which either has the room,
Outside or Communal. That would get rid of the awkward repeated 'room'
key. Or maybe this is OK for now?

dgmstuart committed:
(refactor) Use dig

In most cases it's a bug if an answer hash key doesn't exist, so we want
an exception. However, in cases where it's genuinely optional, dig seems
to be a better way than what we're previously doing: it searches for
nested keys and returns nil if any are missing.

leeky commented on app/models/question_set.rb:
```diff
@@ -14,7 +14,7 @@ def find(id)
     end
 
-    Question.new(@questions.fetch(id))
+    Question.new(id, @questions.fetch(id))
```
How about adding the `id` key to the `question_hash` when it gets loaded instead? (I'm not sure if I like how `id` need to be referenced twice in the initialiser)

```ruby
questions.each do |question, answers|
  [question, answers.merge!('id' => question)]
end
```

Also means you don't have to change any of the tests.

leeky commented on app/models/questions_validator.rb:
```diff
@@ -30,6 +32,17 @@ def validate_desc
     raise MissingPartials, missing_partials if missing_partials.any?
   end
 
+  def validate_mandatory
```
:+1:

leeky commented on app/models/question_set.rb:
```diff
@@ -26,4 +16,32 @@ def find(id)
+  class QuestionSetLoader
+    def initialize(partial_checker:)
+      @partial_checker = partial_checker
+      @root_url = Rails.root
```
`@root_url` doesn't appear to be used anywhere..? As an aside, `Rails.root` is the absolute path to the root of the app rather than a URL.

dgmstuart replied:
Oops! that was left over from the refactoring. 

Thanks for the aside! 👍 

Have amended this commit

leeky commented on app/services/question_saver.rb:
```diff
@@ -16,6 +16,19 @@ def save(form)
   private
 
   def persist_answers(form)
+    persist_room(form)
```
I'm wondering if these 'special' answers should be stored in a separate place?

`SelectedAnswerStore` is now storing: 1) the answers given to questions by a user *AND* 2) specific answers + sor code that are sent to the API when 'creating' a repair, which feel like different things to me.

dgmstuart replied:
I agree, but I'm not sure what the improved solution would look like yet. 

I feel like what we've got now doesn't actually have any problems yet, so it's good enough for now, but I think we should keep an eye open for problems and refactoring to avoid them when we find them.

leeky commented on app/models/repair_params.rb:
```diff
@@ -4,9 +4,9 @@ def initialize(answers)
-    @answers.fetch('describe_repair').fetch('description').tap do |problem|
-      return 'n/a' if problem.blank?
-    end
+    return 'n/a' if description.blank?
```
Won't this cause the room information to be lost when description is blank..?

dgmstuart replied:
Good catch! I'll add a spec

leeky commented on app/presenters/confirmation.rb:
```diff
@@ -43,7 +43,7 @@ def scheduled_action
 
   def room
-    @answers.fetch('room').fetch('room')
+    @answers.dig('room', 'room')
```
I'd forgotten about `dig`... :+1:

dgmstuart replied:
Me too!

dgmstuart committed:
(refactor) Pass question Id in as part of the hash

This does feel slightly better (fewer parameters is usually better), but I don't
think we get away without updating the tests because otherwise we're
using unrealistic test data.

I think the solution to DRYing this up would be to introduce a factory
for creating questions (I don't mean FactoryBot, just a helper method),
but I'd probably only want to use that for the feature specs anyway.

dgmstuart committed:
(fix) Ensure room is passed if no description

This is kind of the whole point of this feature!

I wish there was a better way to do this if statement, but I don't know
of one.

dgmstuart commented:
Rebased and addressed comments - @leeky: should be good for another look

leeky commented on spec/features/users_can_diagnose_their_issue_spec.rb:
```diff
@@ -26,6 +26,7 @@
       .with('first')
       .and_return(
         Question.new(
+          'id' => 'fun',
```
So that the stub matches reality, shouldn't `id` be `first` in this case?

dgmstuart replied:
Yup

leeky commented on spec/features/users_can_diagnose_their_issue_spec.rb:
```diff
@@ -43,6 +44,7 @@
       .with('second')
       .and_return(
         Question.new(
+          'id' => 'sure',
```
Same here - `id` should be `second`, to match the id that `#find` was sent.

leeky commented on spec/fixtures/basic_question.yml:
```diff
@@ -1,4 +1,15 @@
 ---
+location:
```
Is it worth splitting this out into a new fixture? It now contains *three* questions, so is no longer just a 'basic question'.

dgmstuart replied:
I could remove the final question, and then this is the smallest valid question?

leeky replied:
That would be better :+1:

leeky replied:
Ideally, it should be possible to overwrite the mandatory questions array in tests, so that the basic question is still valid. However, I'm not going to worry about this for now.

leeky commented:
@dgmstuart As I'd only suggested minor changes, I've gone ahead and added these myself, in order that I can merge is in your absence.
